### PR TITLE
Add API for creating lambda expressions

### LINKED ISSUE #416: Generating multiline lambda function
``` 
val unknownFields = reader.decodeMessage { tag ->
                when (tag) {
                    1 -> name = ProtoAdapter.STRING.decode(reader)
                    2 -> id = ProtoAdapter.INT32.decode(reader)
                    3 -> email = ProtoAdapter.STRING.decode(reader)
                    4 -> phone.add(PhoneNumber.ADAPTER.decode(reader))
                    else -> UnkownFieldsBuilder.UNKNOWN_FIELD
                }
            }
```

What would be the best way to generate this ? Can we use a combination of begin / end control flow statements ?

### PR DESCRIPTION
**Pull request by Egorand**
Fixes #416 

cc @mc-lovin

### CODE REVIEW

File: src/main/java/com/squareup/kotlinpoet/CodeBlock.kt [hunk 1]
```diff
@@ -354,10 +354,44 @@ class CodeBlock private constructor(
...
+     * @param controlFlow the control flow construct and its code, such as "if (foo == 5)".
+     *     Shouldn't contain braces or newline characters.
+     * @param parameterNames optional control flow parameter names. Allow producing expressions such
+     * as:
+     *
+     * ```
+     * list.forEachIndexed { index, value ->
+     * ```
+     */
+    fun beginParameterizedControlFlow(controlFlow: String, vararg parameterNames: String) =
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/CodeBlock.kt]: This seems like a weird name to describe what always seems to be a lambda. Is it ever not a lambda? I guess you could argue that it's control flow-like, but it seems like a stretch of the term. And then my other concern is that "parameterized" is a term that's usually associated with generics.

Comment (Egorand) [reply src/main/java/com/squareup/kotlinpoet/CodeBlock.kt]: Good point, the naming is off here. I can try changing it to `beginLambdaExpression` and tinker with param names to see what works better.

File: src/main/java/com/squareup/kotlinpoet/CodeBlock.kt [hunk 2]
```diff
@@ -354,10 +354,34 @@ class CodeBlock private constructor(
...
+    fun beginLambdaExpression(statement: String, vararg parameters: String) =
+        beginLambdaExpression(CodeBlock.of(statement), *parameters)
+
+    /**
+     * @param statement statement that precedes the lambda expression, e.g. a function call that
+     * takes the expression as an argument.
+     * @param parameters optional lambda expression parameter names.
+     */
+    fun beginLambdaExpression(
+      statement: CodeBlock,
```
Comment (swankjesse) [inline src/main/java/com/squareup/kotlinpoet/CodeBlock.kt]: sort of a weird word because it’s not a statement

Comment (swankjesse) [reply src/main/java/com/squareup/kotlinpoet/CodeBlock.kt]: codeBlock

**Pull request closed without merging**